claripy/backends/backend_smtlib_solvers/__init__.py

```diff
@@ -0,0 +1,231 @@
+                f.write(smt_script.encode())
+
+        solver.reset()
+        solver.write(smt_script)
+
+        sat = solver.read_sat()
+        if sat == 'sat':
+            model_string = solver.read_model()
+            tokens = Tokenizer(cStringIO(model_string), interactive=True)
+            ass_list = SMTParser(tokens).consume_assignment_list()
+            return sat, {s: val for s, val in ass_list}, ass_list
+        else:
+            error = solver.readline()
+
+        return sat, error, None
+
+    def _eval(self, expr, n, extra_constraints=(), solver=None, model_callback=None):
+        e_c = list(extra_constraints)
+
+        if expr.is_constant():
+            return [expr.constant_value()]
+
+        expr_vars = expr.get_free_variables()
+
+        results = []
+        while len(results) < n:
+            sat, model, ass_list = self._get_model(solver=solver, extra_constraints=e_c, extra_variables=expr_vars)
+            if sat != 'sat':
+                break
+
+            val = self._get_primitive_for_expr(model, expr)
+
+            if val in results:
+                # import ipdb; ipdb.set_trace()
+                raise ValueError("Solver error, solver returned the same value twice incorrectly!")
+
+            results.append(val)
+            # e_c.append(And(*[NotEquals(s, val) for s, val in ass_list]))
+            e_c.append(NotEquals(make_pysmt_const_from_type(val, expr.get_type()), expr))
+
+        return tuple(results)
+
+    def eval(self, expr, n, extra_constraints=(), solver=None, model_callback=None):
+        """
+        This function returns up to `n` possible solutions for expression `expr`.
+
+        :param expr: expression (an AST) to evaluate
+        :param n: number of results to return
+        :param solver: a solver object, native to the backend, to assist in
+                       the evaluation (for example, a z3.Solver)
+        :param extra_constraints: extra constraints (as ASTs) to add to the solver for this solve
+        :param model_callback:      a function that will be executed with recovered models (if any)
+        :return:              A sequence of up to n results (backend objects)
+        """
+        if self._solver_required and solver is None:
+            raise BackendError("%s requires a solver for evaluation" % self.__class__.__name__)
+
+        results = self._eval(
+            self.convert(expr), n, extra_constraints=self.convert_list(extra_constraints),
+            solver=solver, model_callback=model_callback
+        )
+
+        results = list(results)
+        if type(expr) is not BV:
+            return results
+
+        size = expr.length
+        for i in range(len(results)):
+            results[i] &= (1 << size) - 1 # convert it back to unsigned
+
+        # solver is done, terminate process
+        solver.terminate()
+
+        return results
+
+from . import cvc4_popen
+from . import z3_popen
+from . import abc_popen
+from . import z3str_popen
```

Comment:
A new line is required at the end of this file.

Reply:
Fixed now